File: src/javascript/algorithms/strings/minOperationsToEqualFrequencies/minOperationsToEqualFrequencies.js
// Approach: Frequency Counting + Try All Target Frequencies TC: O(n) SC: O(1)
function minOpsToEqualFrequencies(s) {
    const freqCount = new Array(26).fill(0);
    for (const ch of s) {
        freqCount[ch.charCodeAt(0) - 'a'.charCodeAt(0)]++;
    }

    const maxFreq = Math.max(...freqCount);
    let ans = Infinity;

    for (let target = 1; target <= maxFreq; target++) {
        let ops = 0;
        for (const freq of freqCount) {
            if (freq === 0) continue;
            if (freq > target) {
                ops += freq - target;   // delete excess
            } else if (freq < target) {
                ops += freq;            // delete all occurrences
            }
            // freq === target: 0 ops needed
        }
        ans = Math.min(ans, ops);
    }

    return ans;
}

function runTests() {
    const tests = [
        { input: "abc",       expected: 0 },
        { input: "aab",       expected: 1 },
        { input: "aabbc",     expected: 1 },
        { input: "aaabbc",    expected: 2 },
        { input: "aaabbbccc", expected: 0 },
    ];

    for (const { input, expected } of tests) {
        const result = minOpsToEqualFrequencies(input);
        const status = result === expected ? "[PASS]" : "[FAIL]";
        console.log(`${status} input="${input}" result=${result} expected=${expected}`);
    }
}

runTests();

File: src/python/algorithms/strings/minOperationsToEqualFrequencies/min_operations_to_equal_frequencies.py
def min_ops_to_equal_frequencies(s: str) -> int:
    # Approach: Frequency Counting + Try All Target Frequencies TC: O(n) SC: O(1)
    freq_count = [0] * 26
    for ch in s:
        freq_count[ord(ch) - ord('a')] += 1

    max_freq = max(freq_count)
    ans = float('inf')

    for target in range(1, max_freq + 1):
        ops = 0
        for freq in freq_count:
            if freq == 0:
                continue
            if freq > target:
                ops += freq - target    # delete excess
            elif freq < target:
                ops += freq             # delete all occurrences
            # freq == target: 0 ops needed
        ans = min(ans, ops)

    return ans


if __name__ == "__main__":
    tests = [
        ("abc",       0),
        ("aab",       1),
        ("aabbc",     1),
        ("aaabbc",    2),
        ("aaabbbccc", 0),
    ]

    for s, expected in tests:
        result = min_ops_to_equal_frequencies(s)
        status = "[PASS]" if result == expected else "[FAIL]"
        print(f"{status} input=\"{s}\" result={result} expected={expected}")
